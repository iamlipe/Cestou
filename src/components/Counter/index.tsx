import React, {memo, useEffect, useState} from 'react';
import styled, {useTheme} from 'styled-components/native';
import Icon from 'react-native-vector-icons/Feather';
import {useController} from 'react-hook-form';

interface Props {
  name: string;
  control: any;
  maxQuantity?: number;
}

const Counter = ({maxQuantity, name, control}: Props) => {
  const [counter, setCounter] = useState(maxQuantity || 0);
  const theme = useTheme();

  const {
    field: {onChange},
  } = useController({name, control, defaultValue: counter});

  useEffect(() => {
    if (maxQuantity) {
      onChange(maxQuantity);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <StyledContainer>
      <StyledButton>
        <Icon
          testID="decrement-button"
          name="minus"
          color={theme.colors.GRAY_900}
          size={14}
          onPress={() => {
            const newCounter = counter - 1 < 0 ? 0 : counter - 1;

            setCounter(newCounter);
            onChange(newCounter);
          }}
        />
      </StyledButton>
      <StyledInputCounter
        testID={`counter-${name}`}
        value={String(counter)}
        editable={false}
      />
      <StyledButton>
        <Icon
          testID="increment-button"
          name="plus"
          color={theme.colors.GRAY_900}
          size={14}
          onPress={() => {
            if (maxQuantity) {
              const newCounter =
                counter + 1 > maxQuantity ? maxQuantity : counter + 1;

              setCounter(newCounter);
              onChange(newCounter);
            } else {
              const newCounter = counter + 1;

              setCounter(newCounter);
              onChange(newCounter);
            }
          }}
        />
      </StyledButton>
    </StyledContainer>
  );
};

const StyledContainer = styled.View`
  height: 32px;
  width: 86px;
  flex-direction: row;
  align-self: center;
  justify-content: space-between;
  align-items: center;
  background-color: ${({theme}) => theme.colors.PRIMARY_100};
  border-radius: 16px;
  border: 1px solid ${({theme}) => theme.colors.PRIMARY_500};
  padding: 0px 10px;
`;

const StyledInputCounter = styled.TextInput`
  width: 10px;
  font-family: ${({theme}) => theme.fonts.REGULAR_SOURCESANSPRO};
  font-size: ${({theme}) => theme.sizing.SMALLER};
  color: ${({theme}) => theme.colors.GRAY_900};
  padding: 0;
`;

const StyledButton = styled.TouchableOpacity``;

export default memo(Counter);
