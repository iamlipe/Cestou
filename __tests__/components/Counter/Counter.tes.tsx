import React from 'react';
import {
  renderWithThemeProvider,
  matchSnapshotWithProvider,
} from '@/config/functions/Testing';

import Counter from '@/components/Counter';
import {fireEvent} from '@testing-library/react-native';

jest.mock('react-hook-form', () => ({
  useController: () => ({
    field: {
      onChange: jest.fn(),
    },
  }),
}));

describe('Counter', () => {
  const control = jest.fn();
  const name = 'test';

  test('should render correctly', () => {
    renderWithThemeProvider(<Counter name={name} control={control} />);
  });

  test('should matches snapshot', () => {
    matchSnapshotWithProvider(<Counter name={name} control={control} />);
  });

  test('should render with max quantity correctly', () => {
    const counter = renderWithThemeProvider(
      <Counter name={name} control={control} maxQuantity={5} />,
    );

    const inputCounter = counter.getByTestId('counter-test');

    expect(inputCounter.props.value).toEqual('5');
  });

  test('should change value input to more one when press increment button', () => {
    const counter = renderWithThemeProvider(
      <Counter name={name} control={control} />,
    );

    const incrementButton = counter.getByTestId('increment-button');

    fireEvent.press(incrementButton);

    const inputCounter = counter.getByTestId('counter-test');

    expect(inputCounter.props.value).toEqual('1');
  });

  test('should change value input to minus one when press decrement button', () => {
    const counter = renderWithThemeProvider(
      <Counter name={name} control={control} maxQuantity={3} />,
    );

    const decrementButton = counter.getByTestId('decrement-button');

    fireEvent.press(decrementButton);

    const inputCounter = counter.getByTestId('counter-test');

    expect(inputCounter.props.value).toEqual('2');
  });

  test('should not increment value when the new value is bigger then maxQuantity', () => {
    const counter = renderWithThemeProvider(
      <Counter name={name} control={control} maxQuantity={3} />,
    );

    const incrementButton = counter.getByTestId('increment-button');

    fireEvent.press(incrementButton);

    const inputCounter = counter.getByTestId('counter-test');

    expect(inputCounter.props.value).toEqual('3');
  });

  test('should not decrement value when the new value is less then 0', () => {
    const counter = renderWithThemeProvider(
      <Counter name={name} control={control} />,
    );

    const decrementButton = counter.getByTestId('decrement-button');

    fireEvent.press(decrementButton);

    const inputCounter = counter.getByTestId('counter-test');

    expect(inputCounter.props.value).toEqual('0');
  });
});
