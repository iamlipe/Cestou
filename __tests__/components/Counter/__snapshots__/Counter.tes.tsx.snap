// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter should matches snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#C1E9CF",
        "borderColor": "#00B761",
        "borderRadius": 16,
        "borderStyle": "solid",
        "borderWidth": 1,
        "flexDirection": "row",
        "height": 32,
        "justifyContent": "space-between",
        "paddingBottom": 0,
        "paddingLeft": 10,
        "paddingRight": 10,
        "paddingTop": 0,
        "width": 86,
      },
    ]
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <Text
      allowFontScaling={false}
      onPress={[Function]}
      selectable={false}
      style={
        Array [
          Object {
            "color": "#1A1A1A",
            "fontSize": 14,
          },
          undefined,
          Object {
            "fontFamily": "Feather",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
      testID="decrement-button"
    >
      
    </Text>
  </View>
  <TextInput
    editable={false}
    style={
      Array [
        Object {
          "color": "#1A1A1A",
          "fontFamily": "SourceSansPro-Regular",
          "fontSize": 27,
          "paddingBottom": 0,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 0,
          "width": 10,
        },
      ]
    }
    testID="counter-test"
    value="0"
  />
  <View
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <Text
      allowFontScaling={false}
      onPress={[Function]}
      selectable={false}
      style={
        Array [
          Object {
            "color": "#1A1A1A",
            "fontSize": 14,
          },
          undefined,
          Object {
            "fontFamily": "Feather",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
      testID="increment-button"
    >
      
    </Text>
  </View>
</View>
`;
